### ISSUE TITLE
[ RFC ] Improved modules API

### ISSUE THREAD

**Issue by filrak** [DivanteLtd/vue-storefront#2760]
## What is the motivation for adding / enhancing this feature?

Current modules API is a huge milestone in VS history but still pretty limited. The goal of this transformation is to cut off those limits.

The biggest limitations are
- inability to load modules lazily in easy way
- modules merging during runtime process (old and new module are landing in final bundle) - still need to figure it out and it's probably a case for another issue

The goal of the solution is to solve those two problems while not changing any currently existing module from the inside. In other words this change should be handled purely on Vue Storefront level so there are no changes needed in any of currently existing modules.

## Solution

To allow lazy loading of modules and better error handling we should put an abstraction layer on top of Vuex (similar to router manager). I was trying to do it seamlessly as a proxy ([see reference](https://github.com/DivanteLtd/vue-storefront/pull/2164)) but it turned out to be impossible. To use any of the modules the module object needs to be called directly which should enable seamless lazy loading,  autocompletion and better error handling
```js
//somewhere in code
const CatalogModule = () => import('@vue-storefront/catalog')
CatalogModule.dispatch('actionName', payload)
```

We will also depreciate old root registration in `index.ts` and `extendModule` method in favor of `VueStorefrontModule.extend()` so everything is encapsulated.

This is something I would like to do during Christmas break. 

Ofc everything will be backward comaptible and for next 3 major versions you could still use old way of registering and using modules.

**Comment by lukeromanowicz** [comment]
While we are here, would be nice to add dependency checking between vsf modules. I.e. if there is a checkout module, build process should check if the cart module is being loaded as well. We would probably want also to have control over which modules load first based on something else than alphabetic order.

**Comment by filrak** [comment]
@lukeromanowicz the order will be specified in the array, regarding dependency checking I guess it will be handled in the same way as in theme:

If we need module B in module A we just use a dynamic import inside module A and syntax as suggested in RFC. Then every dependency will be downloaded only while needed

**Comment by pkarw** [comment]
This is awesome. How about adding the module hooks by this chance? I mean hooks like: afterApplicationLoaded, afterStoreviewChanged (there were few cases we encountered with @patzick thay there’s is no entry point for applying some business logic once the proper store view is loaded)? I think it would be great to use similar approach like Nuxt has and set the fixed set of endpoints; current set is too limited I belive

**Comment by filrak** [comment]
Seems like a good idea :) I guess THIS is what event bus should be used for, not for stuff like before/after filter changed etc ;)

**Comment by pkarw** [comment]
@filrak yeah, the hooks are better because they’re easier to understand/ discover for new comers - the hell has just frozen as I’m saying that hooks are >in some cases<  better than events haha :) go on and propose the solution, you’re our module master 👑 ;)

**Comment by filrak** [comment]
The question is - should those hooks be fired only AFTER the module is loaded (listeners are setted up when module is loaded by the browser) or ALWAYS no matter if module was loaded or not.
 keep in mind that some module may be loaded only on certain situations.

If the answer is AFTER module is loaded then we can even omit the config file registration since it won't do anything important. It seems like a proper approach since otherwise we would need to split module into the part that is loaded right after app init (hooks) and the rest that is loaded lazily (vuex etc).

**Comment by pkarw** [comment]
That’s good point. Some modules are and will be always loaded and there won’t be a problem. The others should probably use something different than hooks (?).

I mean we shouldn’t call the hooks of not-loaded modules I guess (?)

**Comment by filrak** [comment]
Yes we couldn't but I was trying to figure out if this is even realistic scenario and I guess it isn't. If a module needs to be loaded it needs to loaded fully. If we want to hook into initialization parts of the app tha module should be loaded during intiialization. Loading only the before/after makes no sense since inside of them we are usually refering to internals of module (like vuex actions).

I guess the rubber duck method led us to the conclusion that the only thing we need is abstraction over Vuex when it comes to modules lazy loading. 

Extending them though is much more complex problem than I initially thought when it comes to dead code elimination. Still need to figure this part out. I will come up with a solution during Easter break ^^

**Comment by filrak** [comment]
Closed in favor to new proposal

**Comment by kkdg** [comment]
I have no idea why github shows that `I added this to the 2.0 milestone` when I didn't do anything. hmm..
